src/logcollector/state.c

```diff
@@ -0,0 +1,378 @@
+#else
+#define STATIC static
+#endif
+
+#define W_LC_STATE_TIME_FORMAT "%Y-%m-%d %H:%M:%S" ///< Time format for the JSON and the file output
+#define W_LC_STATE_TIME_LENGHT (19 + 1)            ///< Maximum time size
+
+/* Global variables */
+
+w_lc_state_type_t g_lc_state_type;       ///< state enabled flag
+bool g_lc_state_file_enabled;          ///< state file dump enable
+cJSON * g_lc_json_stats;               ///< JSON representation of states
+w_lc_state_storage_t * g_lc_states_global;      ///< global state struct storage
+w_lc_state_storage_t * g_lc_states_interval;    ///< interval state struct storage
+pthread_mutex_t g_lc_raw_stats_mutex;  ///< g_lc_states_* structs mutual exclusion mechanism
+pthread_mutex_t g_lc_json_stats_mutex; ///< g_lc_json_stats mutual exclusion mechanism
+
+/**
+ * @brief Trigger the generation of states
+ *
+ */
+STATIC void w_logcollector_state_generate();
+
+/**
+ * @brief Generate and process the current states information
+ *
+ * @param state state to generate
+ * @param restart restart counters and date after generating
+ * @return cJSON * json decription with state information
+ */
+STATIC cJSON * _w_logcollector_generate_state(w_lc_state_storage_t * state, bool restart);
+
+/**
+ * @brief Update/register current event and byte count for a particular file/location
+ *
+ * @param state state to be used
+ * @param fpath file path or locafile location value
+ * @param bytes amount of bytes
+ */
+STATIC void _w_logcollector_state_update_file(w_lc_state_storage_t * state, char * fpath, uint64_t bytes);
+
+/**
+ * @brief Update/register current drop count for a target belonging to a particular file
+ *
+ * @param state state to be used
+ * @param fpath file path or locafile location value
+ * @param target target name
+ * @param dropped true if want to register a drop.
+ */
+STATIC void _w_logcollector_state_update_target(w_lc_state_storage_t * state, char * fpath, char * target, bool dropped);
+
+/**
+ * @brief Dump state information to file
+ *
+ */
+STATIC void w_logcollector_state_dump();
+
+#ifdef WIN32
+DWORD WINAPI w_logcollector_state_main(void * args) {
+#else
+void * w_logcollector_state_main(void * args) {
+#endif
+
+    int interval = *(int *) args;
+
+    if (interval > 0) {
+        while (FOREVER()) {
+            sleep(interval);
+            w_logcollector_state_generate();
+            if (g_lc_state_file_enabled) {
+                w_logcollector_state_dump();
+            }
+        }
+    }
+
+#ifndef WIN32
+    return NULL;
+#else
+    return 0;
+#endif
+}
+
+STATIC void w_logcollector_state_dump() {
+
+    cJSON * lc_state_json = w_logcollector_state_get();
+    char * lc_state_str = cJSON_Print(lc_state_json);
+    cJSON_Delete(lc_state_json);
+
+    // Add trailing newline
+    const size_t len = strlen(lc_state_str);
+    os_realloc(lc_state_str, len + 2, lc_state_str);
+    lc_state_str[len] = '\n';
+    lc_state_str[len + 1] = '\0';
+
+    FILE * lc_state_file = NULL;
```

Comment:
I would assign the **_fopen_**'s output here and would add the "**_const_**" modifier